reconcile/utils/oc_map.py

```diff
@@ -0,0 +1,317 @@
+                threaded.run(
+                    self.init_oc_client,
+                    clusters_dict.values(),
+                    self._thread_pool_size,
+                    privileged=False,
+                )
+            if privileged_clusters:
+                threaded.run(
+                    self.init_oc_client,
+                    privileged_clusters.values(),
+                    self._thread_pool_size,
+                    privileged=True,
+                )
+        else:
+            raise KeyError("expected one of clusters or namespaces.")
+
+    def set_jh_ports(self, jh: MutableMapping[Any, Any]) -> None:
+        # This will be replaced with getting the data from app-interface in
+        # a future PR.
+        jh["remotePort"] = 8888
+        key = f"{jh['hostname']}:{jh['remotePort']}"
+        with self._lock:
+            if key not in self._jh_ports:
+                port = JumpHostSSH.get_unique_random_port()
+                self._jh_ports[key] = port
+            jh["localPort"] = self._jh_ports[key]
+
+    def init_oc_client(self, cluster_info: HasCluster, privileged: bool) -> None:
+        cluster = cluster_info.name
+        if not privileged and self._oc_map.get(cluster):
+            return None
+        if privileged and self._privileged_oc_map.get(cluster):
+            return None
+        if self.cluster_disabled(cluster_info):
+            return None
+        if self._internal is not None:
+            # integration is executed with `--internal` or `--external`
+            # filter out non matching clusters
+            if self._internal and not cluster_info.internal:
+                return
+            if not self._internal and cluster_info.internal:
+                return
+
+        if privileged:
+            automation_token = cluster_info.cluster_admin_automation_token
+            token_name = "clusterAdminAutomationToken"
+        else:
+            automation_token = cluster_info.automation_token
+            token_name = "automationToken"
+
+        if automation_token is None:
+            self.set_oc(
+                cluster,
+                OCLogMsg(
+                    log_level=logging.ERROR, message=f"[{cluster}] has no {token_name}"
+                ),
+                privileged,
+            )
+        # serverUrl isn't set when a new cluster is initially created.
+        elif not cluster_info.server_url:
+            self.set_oc(
+                cluster,
+                OCLogMsg(
+                    log_level=logging.ERROR, message=f"[{cluster}] has no serverUrl"
+                ),
+                privileged,
+            )
+        else:
+            server_url = cluster_info.server_url
+            insecure_skip_tls_verify = cluster_info.insecure_skip_tls_verify
+
+            try:
+                if not self._secret_reader:
+                    raise Exception("No secret_reader set")
+                token = self._secret_reader.read_secret(automation_token)
```

Comment:
it would be great if we would not need the secret reader in `Oc_Map`. you have more context right now so wdyt about resolving secrets before handing clusters to `OC_Map`.

this question is clearly out of scope for this PR. carry to thought over to the followup PRs if you like

Reply:
I like the idea a lot. Once the underlying classes `OCDeprecated`, `OCNative` and `SSHJumphost` are type compatible, I will try to remove the need of secret reader. It makes sense :+1: